clang/lib/Driver/ToolChains/Darwin.cpp

```diff
@@ -2539,6 +2550,18 @@ void DarwinClang::AddClangSystemIncludeArgs(const llvm::opt::ArgList &DriverArgs
     llvm::sys::path::append(P, "usr", "include");
     addExternCSystemInclude(DriverArgs, CC1Args, P.str());
   }
+
+  // Add default framework search paths
+  auto addFrameworkInclude = [&](auto ...Path) {
+    SmallString<128> P(Sysroot);
+    llvm::sys::path::append(P, Path...);
+
+    CC1Args.push_back("-internal-iframework");
+    CC1Args.push_back(DriverArgs.MakeArgString(P));
+  };
+  addFrameworkInclude("System", "Library", "Frameworks");
```

Comment:
Do you think it's possible to have the same string input style for linker search paths as header search paths? It would make sharing common the paths simpler, as a future cleanup.

Reply:
Do you mean something like this?

```c++
addFrameworkInclude(DriverArgs, CC1Args, P);
```

That's possible, in fact I started with that. However, I will have to do this:

```c++
auto addFrameworkInclude = [](const llvm::opt::ArgList &DriverArgs, llvm::opt::ArgStringList &CC1Args, SmallString<128> const& P) {
  CC1Args.push_back("-internal-iframework");
  CC1Args.push_back(DriverArgs.MakeArgString(P));
};

{
  SmallString<128> P(Sysroot);
  llvm::sys::path::append(P, "System", "Library", "Frameworks");
  addFrameworkInclude(DriverArgs, CC1Args, P);
}
{
  SmallString<128> P(Sysroot);
  llvm::sys::path::append(P, "System", "Library", "SubFrameworks");
  addFrameworkInclude(DriverArgs, CC1Args, P);  
}
{
  SmallString<128> P(Sysroot);
  llvm::sys::path::append(P, "Library", "Frameworks");
  addFrameworkInclude(DriverArgs, CC1Args, P);  
}
```

Minor variations on this may be possible, but I thought that was less readable. If you have a preference for this (perhaps motivated for future refactoring potential), I'm fine with going for that, just let me know.

Reply:
Could we change the lambda `AddSearchPath` used for linker search args as a static function and reuse it for both places? 
That way we could do 
```
// For header search
addSearchPath(DriverArgs, CC1Args, Sysroot, "-internal-iframework", "/System/Library/Frameworks");

// For linker
addSearchPath(Args, CmdArgs, Sysroot, "-F", "/System/Library/Frameworks");
```

If that's not easily doable, having the `addFrameworkInclude` lambda accept the same types as `AddSearchPath` is nice.